cloud: add n8n doc (#11594)
pingcap/docs: tidb-cloud/integrate-tidbcloud-with-n8n.md

---
title: Integrate TiDB Cloud with n8n

> summary: Learn the use TiDB Cloud node in n8n.

Reviewer: ```suggestion
summary: Learn the use of TiDB Cloud node in n8n.
```

---

# Integrate TiDB Cloud with n8n

[n8n](https://n8n.io/) is an extendable workflow automation tool. With a [fair-code](https://faircode.io/) distribution model, n8n will always have visible source code, be available to self-host, and allow you to add your custom functions, logic ,and apps. 

> This document introduces how to build an auto-workflow, create a TiDB Serverless cluster and write some data into TiDB.

Reviewer: ```suggestion
This document introduces how to build an auto-workflow, create a TiDB Serverless Tier cluster and write some data into TiDB.
```

---
title: Integrate TiDB Cloud with n8n
summary: Learn the use of TiDB Cloud node in n8n.
---

# Integrate TiDB Cloud with n8n

> [n8n](https://n8n.io/) is an extendable workflow automation tool. With a [fair-code](https://faircode.io/) distribution model, n8n will always have visible source code, be available to self-host, and allow you to add your custom functions, logic ,and apps. 

Reviewer 2: ```suggestion
[n8n](https://n8n.io/) is an extendable workflow automation tool. With a [fair-code](https://faircode.io/) distribution model, n8n will always have visible source code, be available to self-host, and allow you to add your custom functions, logic, and apps. 
```

> This document introduces how to build an auto-workflow: create a TiDB Cloud Serverless Tier cluster, gather Hack News RSS, store it to TiDB and send a briefing email.

Reviewer 2: ![image](https://user-images.githubusercontent.com/3690895/208296229-9842b2fe-9ce5-4a71-bd49-f817f068b49e.png)
Hacker News or Hack News?

Author: @zhangyangyu Thank you very much. I will correct it.

## Prerequisites: Get TiDB Cloud API Key

1. Access your TiDB Cloud dashboard.
2. Click on the **account** tab in the top right.
3. Click **Organization Settings**.
4. Click **API Keys** tab.
5. Click on the **Create API Key** button to create a new API Key.
6. Use these **API Keys** with your TiDB Cloud node credentials in n8n.

For more information, see [TiDB Cloud API Overview](https://docs.pingcap.com/tidbcloud/api-overview/).

## Step 1: Install n8n

There are two ways to install your self-hosting n8n, choose whichever works for you.

- **Install by npm**
  1. Install [node.js](https://nodejs.org/en/download/) on your workspace.
  2. Download and start n8n by `npx`.
     ```shell
     npx n8n
     ```
     
- **Install by docker**
  1. Install [Docker](https://www.docker.com/products/docker-desktop) on your workspace.
  2. Download and start n8n by `docker`
     ```shell
     docker run -it --rm --name n8n -p 5678:5678 -v ~/.n8n:/home/node/.n8n n8nio/n8n
     ```

> After starting n8n, you can visit [localhost:5678](http://localhost:5678) to taste n8n.

Reviewer 2: Could consider using tab to display these two.

## Step 2: Install TiDB Cloud Node in n8n

TiDB Cloud node names `n8n-nodes-tidb-cloud` in npm repository. You need to install this node manually to control TiDB Cloud with n8n.

> 1. Sign up [self-hosting n8n](http://localhost:5678).

Reviewer 2: hmm, a localhost link, not sure.

1. Navigate to **Workflows** panel, and click **Add workflow**.
2. In new workflow workspace, click **+** in the top right corner and choose **All** field.
3. Search `TiDB Cloud` and drag it to the workspace.
4. Enter credentials, which is the TiDB Cloud API key, for the TiDB Cloud node.
5. Choose your project from the **Project** list.
6. Select `Create Serverless Cluster` from the **Operation** list.
7. Enter a cluster name in the **Cluster Name** field.
8. Select a region from the **Region** list.
9. Enter a password used to log in to your TiDB clusters in the **Password** field.
10. Click on **Execute Node** to run the node.

> Note: The step takes some 1 minute to create a new TiDB Serverless cluster.

### Create a workflow

#### Use a manual trigger as the workflow's starter

1. If you don't have a workflow. Please navigate to **Workflows** panel, and click **Add workflow**. Otherwise, skip this point.
2. Click **+** in the top right corner and search `schedule trigger`.
3. Drag the manual trigger node to your workspace.
4. Choose `Days` in the **Trigger Interval**.
5. Set **Days Between Triggers** as `1`.

> 6. Choose `8am` in the T**rigger at Hour**.

Reviewer 2: ```suggestion
6. Choose `8am` in the **Trigger at Hour**.
```

```javascript
let message = "";

// Loop the input items
for (item of items) {
  message += `
       <h3>${item.json.title}</h3>
       <br>
       ${item.json.content}
       <br>
       `
}

let response =
    `
       <!DOCTYPE html> 
       <html> 
       <head> 
       <title>Hack News Briefing</title> 
    </head> 
    <body>  
        ${message} 
    </body> 
    </html>
    `
// Return our message
return [{json: {response}}];
```

#### Send message by Gmail

1. Click **+** to the right of the code node.
2. Search `gmail` and add it to the workspace.
3. Enter credentials for the Gmail node, you can find out how to do that [here](https://docs.n8n.io/integrations/builtin/credentials/google/oauth-single-service/).
4. Choose `Message` in the **Resource**. 
5. Choose `Send` in the **Operation**.
6. Enter your email in the **To**.
7. Enter `Hack News Briefing` in the **Subject**.
8. Choose `HTML` in the **Email Type**.
9. Set **Message** field mode to `Expression`
10. Enter `{{ $json["response"] }}` in the **Message**.

> Note: It is very important to mouse over the Message field and select the Expression pattern.

## Step 4: Run Your Workflow

> After building up the workflow, you can click the **Execute Workflow** button to test run it. You'll get Hack News briefing emails ,and these news will be logged to your TiDB Cloud Serverless Tier. So don't be worry about losing them.

Reviewer 2: ```suggestion
After building up the workflow, you can click the **Execute Workflow** button to test run it. You'll get Hack News briefing emails, and these news will be logged to your TiDB Cloud Serverless Tier cluster. So don't worry about losing them.
```

Now you can activate this workflow in the **Workflows** panel. This workflow will help you get the first page articles on Hacker News every day.

## TiDB Cloud Node Core

### Support Operation

TiDB Cloud Node acts as a [regular node](https://docs.n8n.io/workflows/nodes/#regular-nodes), and only supports the following five operations.

- **Create Serverless Cluster**: Create a TiDB Cloud Serverless Tier cluster.
- **Execute SQL**: Execute an SQL statement in TiDB.
- **Delete**：Delete rows in TiDB.
- **Insert**: Insert rows in TiDB.
- **Update**: Update rows in TiDB.

### Fields Description

Different operations require different fields to be filled in. The following shows the respective field descriptions according to the operation.

<SimpleTab>
<div label="Create Serverless Cluster">

- **Credential for TiDB Cloud API**: Only supports TiDB Cloud API key authentication. Refer to [Get TiDB Cloud API Key](#prerequisites-get-tidb-cloud-api-key).
- **Project**: The TiDB Cloud project name. 
- **Operation**: The operation of this node. Refer to [Support Operation](#support-operation).
- **Cluster**: The TiDB Cloud cluster name. Enter one name for your new cluster.

> - **Region**: The region name. Choose a region where your cluster will be deployed. Usually, choose the region closest to your product cluster.

Reviewer 2: ```suggestion
- **Region**: The region name. Choose a region where your cluster will be deployed. Usually, choose the region closest to your application deployment.
```

- **Project**: The TiDB Cloud project name.
- **Operation**: The operation of this node. Refer to [Support Operation](#support-operation).
- **Cluster**: The TiDB Cloud cluster name. In the Create Serverless Cluster operation, you need to enter one name for your new cluster. While in other operations, you should choose one existing cluster.
- **Password**: The password of TiDB Cloud cluster.
- **User**: The username of your TiDB Cloud cluster.
- **Database**: The database name.
- **Table**: The table name. You can use `From list` mode to choose one or `Name` mode to type table manually.
- **Update Key**: The name of the item's property which decides which rows in the database should be updated. Item is the data sent from one node to another. A node performs its action on each item of incoming data. For more information about item in n8n, see [n8n documentation](https://docs.n8n.io/workflows/items/).
- **Columns**: The comma-separated list of the input item which should used as columns for the rows to update.

</div>
</SimpleTab>

### Limitations 
 

> 1. Normally only one SQL statement is allowed in the **Execute SQL** operation. If you want to execute more than one statement in a single operation, you need to manually enable [`tidb_multi_statement_mode`](https://docs.pingcap.com/tidb/dev/system-variables).

Reviewer 2: ```suggestion
1. Normally only one SQL statement is allowed in the **Execute SQL** operation. If you want to execute more than one statement in a single operation, you need to manually enable [`tidb_multi_statement_mode`](https://docs.pingcap.com/tidb/dev/system-variables#tidb_multi_statement_mode-new-in-v4011).
```

---
title: Integrate TiDB Cloud with n8n
summary: Learn the use of TiDB Cloud node in n8n.
---

# Integrate TiDB Cloud with n8n

[n8n](https://n8n.io/) is an extendable workflow automation tool. With a [fair-code](https://faircode.io/) distribution model, n8n will always have visible source code, be available to self-host, and allow you to add your custom functions, logic, and apps. 

This document introduces how to build an auto-workflow: create a TiDB Cloud Serverless Tier cluster, gather Hacker News RSS, store it to TiDB and send a briefing email.

## Prerequisites: Get TiDB Cloud API Key

1. Access your TiDB Cloud dashboard.
2. Click on the **account** tab in the top right.
3. Click **Organization Settings**.
4. Click **API Keys** tab.
5. Click on the **Create API Key** button to create a new API Key.

> 6. Use these **API Keys** with your TiDB Cloud node credentials in n8n.

Author: @ran-huang Thanks for your careful review. TIDB UI updates so quickly that it's really hard to guarantee the timeliness.

- [Gmail](https://docs.n8n.io/integrations/builtin/app-nodes/n8n-nodes-base.gmail/)
- [TiDB Cloud node](https://www.npmjs.com/package/n8n-nodes-tidb-cloud)

The final workflow should look like the following image.

![img](/media/tidb-cloud/integration-n8n-workflow-rss.jpg)

### (Optional) Create a TiDB Cloud Serverless Tier cluster

If you haven't got a TiDB Cloud Serverless Tier cluster, you can use this node to create one. Otherwise, feel free to skip this operation.

1. Navigate to **Workflows** panel, and click **Add workflow**.
2. In new workflow workspace, click **+** in the top right corner and choose **All** field.
3. Search `TiDB Cloud` and drag it to the workspace.
4. Enter credentials, which is the TiDB Cloud API key, for the TiDB Cloud node.
5. Choose your project from the **Project** list.
6. Select `Create Serverless Cluster` from the **Operation** list.
7. Enter a cluster name in the **Cluster Name** field.
8. Select a region from the **Region** list.
9. Enter a password used to log in to your TiDB clusters in the **Password** field.
10. Click on **Execute Node** to run the node.

> **Note:** 
> 

> > This step takes some 1 minute to create a new TiDB Serverless cluster.

Reviewer 2: ```suggestion
> This step takes seconds to create a new TiDB Serverless cluster.
```

- **Project**: The TiDB Cloud project name.
- **Operation**: The operation of this node. Refer to [Support Operation](#support-operation).
- **Cluster**: The TiDB Cloud cluster name. In the Create Serverless Cluster operation, you need to enter one name for your new cluster. While in other operations, you should choose one existing cluster.
- **Password**: The password of TiDB Cloud cluster.
- **User**: The username of your TiDB Cloud cluster.
- **Database**: The database name.
- **Table**: The table name. You can use `From list` mode to choose one or `Name` mode to type table manually.
- **Update Key**: The name of the item's property which decides which rows in the database should be updated. Item is the data sent from one node to another. A node performs its action on each item of incoming data. For more information about item in n8n, see [n8n documentation](https://docs.n8n.io/workflows/items/).
- **Columns**: The comma-separated list of the input item which should used as columns for the rows to update.

</div>
</SimpleTab>

### Limitations 
 

> 1. Normally only one SQL statement is allowed in the **Execute SQL** operation. If you want to execute more than one statement in a single operation, you need to manually enable [`tidb_multi_statement_mode`](https://docs.pingcap.com/tidb/dev/system-variables#tidb_multi_statement_mode-new-in-v4011).

Reviewer 3: Use the link in *TiDB Cloud documentation* rather than *TiDB documentation*.
```suggestion
- Normally only one SQL statement is allowed in the **Execute SQL** operation. If you want to execute more than one statement in a single operation, you need to manually enable [`tidb_multi_statement_mode`](https://docs.pingcap.com/tidbcloud/system-variables#tidb_multi_statement_mode-new-in-v4011).
```

> 2. The **Delete** and **Update** operation need to specify one field as a key. For example, the `Delete Key` is set to `id`, which is equivalent to executing `delete from table where id = ${item.id}`. Currently, it only supports specifying one key.

Reviewer 3: In `it only supports specifying one key.`, it's unclear what the **`it`** refers to.

```suggestion
- The **Delete** and **Update** operations need to specify one field as a key. For example, the `Delete Key` is set to `id`, which is equivalent to executing `DELETE FROM table WHERE id = ${item.id}`. Currently, the **Delete** and **Update** operations only support specifying one key.
```

> 3. The **Insert** and **Update** operation need to specify the comma-separated list in the **Columns** field, and the field name must be the same as the input item's property.

Reviewer 3: ```suggestion
- The **Insert** and **Update** operations need to specify the comma-separated list in the **Columns** field, and the field name must be the same as the input item's property.
```

9. Set **Message** field mode to `Expression`
10. Enter `{{ $json["response"] }}` in the **Message**.

> **Note:** 
> 

> > It is very important to mouse over the Message field and select the Expression pattern.

Reviewer 2: hover

Author: @zhangyangyu Thank you for your careful review. I have used `hover` instead of `over`.

## Step 4: Run Your Workflow

After building up the workflow, you can click the **Execute Workflow** button to test run it. You'll get Hacker News briefing emails, and these news will be logged to your TiDB Cloud Serverless Tier cluster. So don't worry about losing them.

Now you can activate this workflow in the **Workflows** panel. This workflow will help you get the first page articles on Hacker News every day.

> ## TiDB Cloud Node 核心

Reviewer 2: Chinese?
